akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMethod.scala

```diff
@@ -32,41 +32,77 @@ object RequestEntityAcceptance {
  * @param requestEntityAcceptance Expected if meaning of request entities is properly defined
  */
 final case class HttpMethod private[http] (
-  override val value:      String,
-  isSafe:                  Boolean,
-  isIdempotent:            Boolean,
-  requestEntityAcceptance: RequestEntityAcceptance) extends jm.HttpMethod with SingletonValueRenderable {
+  override val value:       String,
+  isSafe:                   Boolean,
+  isIdempotent:             Boolean,
+  requestEntityAcceptance:  RequestEntityAcceptance,
+  val contentLengthAllowed: StatusCode => Boolean) extends jm.HttpMethod with SingletonValueRenderable {
   override def isEntityAccepted: Boolean = requestEntityAcceptance.isEntityAccepted
   override def toString: String = s"HttpMethod($value)"
 }
 
 object HttpMethod {
+
+  // the allowsEntity condition was used to determine what responses provided the Content-Length header, before #4213 was fixed
+  private def oldContentLengthCondition(status: StatusCode) = status.allowsEntity
+
+  /**
+   * Create a custom method type.
+   * @deprecated The created method will compute the presence of Content-Length headers based on deprecated logic (before issue #4213).
+   */
+  @Deprecated
   def custom(name: String, safe: Boolean, idempotent: Boolean, requestEntityAcceptance: RequestEntityAcceptance): HttpMethod = {
     require(name.nonEmpty, "value must be non-empty")
     require(!safe || idempotent, "An HTTP method cannot be safe without being idempotent")
-    apply(name, safe, idempotent, requestEntityAcceptance)
+    apply(name, safe, idempotent, requestEntityAcceptance, oldContentLengthCondition)
+  }
+
+  // declare constant functions so custom HttpMethod instances are equal (case class equality) when built with the same parameters
+  private val constantTrue: StatusCode => Boolean = (_) => true
+  private val constantFalse: StatusCode => Boolean = (_) => false
```

Comment:
You can use `akka.util.ConstantFun#anyToTrue` and `akka.util.ConstantFun#anyToFalse` instead of defining these.